prototype_source/aot_inductor_tutorial.py

```diff
@@ -0,0 +1,138 @@
+# run this binary.
+#
+# Note: a modern NVIDIA GPU (H100, A100, or V100) is recommended for this
+# tutorial in order to reproduce the speedup numbers shown below and documented
+# elsewhere.
+
+######################################################################
+# Entrypoint
+# ----------
+#
+# The entrypoint to AOTInductor is through the ``torch._export.aot_compile``
+# function. Note that this function is still in prototype stage and is subject to change.
+#
+# The signature is:
+#
+# .. code:: python
+#
+#   torch._export.aot_compile(
+#       module: Callable,
+#       args: Tuple[Any, ...],
+#       kwargs: Optional[Dict[str, Any]] = None,
+#       *,
+#       dynamic_shapes: Optional[Dict[str, Any]] = None,
+#       options: Optional[Dict[str, Any]] = None,
+#   ) -> Tuple[str, bytes]
+#
+# This function does the following:
+#    1. Exports the given model into a compilation graph through the
+#      `torch.export <https://pytorch.org/docs/2.1/export.html>`__ workflow. If
+#      there are any errors when exporting the model, please refer to the
+#      `torch.export <https://pytorch.org/docs/2.1/export.html>`__ documentation.
+#    2. Compiles the exported program using TorchInductor into a shared library.
+#    3. Returns the path to the shared library.
+#
+
+import torch
+from torch.export import export
+import torch.utils._pytree as pytree
+
+class M(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.branch1 = torch.nn.Sequential(
+            torch.nn.Linear(64, 32), torch.nn.ReLU()
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.branch1(x)
+
+with torch.no_grad():
+    m = M().to("cuda")
+    example_args = (torch.randn(32, 64, device="cuda"),)
+    example_outputs = m(*example_args)
+    _, output_spec = pytree.tree_flatten(example_outputs)
+
+    so_path, exported_program = torch._export.aot_compile(
+        m, args=example_args
+    )
+
+print(".so path:")
+print(so_path)
+print()
+print("Exported graph passed to inductor:")
+print(exported_program)
+print()
+
+######################################################################
+# We can also inspect the generated C++ code associated with the shared library
+# through looking at the ``so_path``:
+
+print("Generated C++ Code:")
+with open(f"{so_path[:-3]}.cpp", 'r') as f:
+    print(f.read())
+print()
+
+######################################################################
+# Running the .so
+# ---------------
+#
+# To run the .so, we can import ``aot_inductor_launch`` from
+# ``torch._inductor.utils`` which includes some starter C++ code to launch the
+# program.
+#
+# The runnable program takes in a list of flattened arguments, and returns a
+# list of flattened outputs, so we will need to do a little bit of preprocessing
+# to match the calling/returning convention of the eager module.
+
+from torch._inductor.utils import aot_inductor_launcher
```

Comment:
I suppose this section could be folded into some utils.

I don't like that you're carrying around `output_spec` from way above where you had to run the model to get it. Is it not available via the `exported_program` itself? If so, why not retrieve it from there?

Then you could basically define a util that takes the `(so_path, exported_program)` output from `aot_compile` and runs it as below, first by loading the container, then running it, then unflattening.